transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -218,28 +239,39 @@ private void doReleaseChannel(Channel channel, Promise<Void> promise) {
                          // Better include a stracktrace here as this is an user error.
                          new IllegalArgumentException(
                                  "Channel " + channel + " was not acquired from this ChannelPool"),
-                         promise);
+                         promise, null);
         } else {
             try {
                 if (offerChannel(channel)) {
                     handler.channelReleased(channel);
                     promise.setSuccess(null);
                 } else {
-                    closeAndFail(channel, FULL_EXCEPTION, promise);
+                    closeAndFail(channel, FULL_EXCEPTION, promise, null);
                 }
             } catch (Throwable cause) {
-                closeAndFail(channel, cause, promise);
+                closeAndFail(channel, cause, promise, null);
             }
         }
     }
 
-    private static void closeChannel(Channel channel) {
+    private static void closeChannel(final Channel channel, final Promise channelClosePromise) {
         channel.attr(POOL_KEY).getAndSet(null);
-        channel.close();
+        ChannelFuture future = channel.close();
+        if (channelClosePromise != null) {
+            future.addListener(new GenericFutureListener<Future<? super Void>>() {
+                @Override
+                public void operationComplete(Future<? super Void> future) throws Exception {
+                    if (future.isSuccess()) {
```

Comment:
Shouldn't we call also "channelClosePromise.setFailure(...)" if not success full ?